model_zoo/mnist/customized_optimizers.py

```diff
@@ -0,0 +1,376 @@
+        the batches on which we communicated the locally aggregated gradients.
+        """
+        flattended_args0 = [item for tup in args[0] for item in tup]
+
+        # Since we skip applying updates when the counter is not at zero we
+        # still want to increment the global step if it is being tracked
+        # (e.g., Tensorflow Estimators).
+        def increment_global_step_counter():
+            global_step_counter = tf.compat.v1.train.get_global_step()
+            if global_step_counter is None:
+                return tf.no_op()
+            return global_step_counter.assign_add(
+                tf.constant(1, dtype=tf.int64),
+                use_locking=True,
+                read_value=False
+            )
+
+        # Increment global step on iterations where we don't call `apply_gradients()`.
+        cond_increment_global_step_counter = tf.cond(
+            pred=tf.equal(self.counter, 0),
+            true_fn=tf.no_op,
+            false_fn=increment_global_step_counter,
+        )
+        flattended_args0.append(cond_increment_global_step_counter)
+
+        # If optimizer tracks iterations, we increment it on steps where we
+        # are not going to call `apply_gradients()`.
+        def increment_optimizer_iteration():
+            if hasattr(optimizer, "_iterations") and optimizer._iterations is not None:
+                return optimizer._iterations.assign_add(1).op
+            return tf.no_op()
+
+        with tf.control_dependencies([tf.group(*get_not_none_from_list(flattended_args0))]):
+            return tf.cond(
+                pred=tf.equal(self.counter, 0),
+                true_fn=apply_grads_closure,
+                false_fn=increment_optimizer_iteration,
+            )
+
+
+## `ElasticDistributedOptimizer` is slightly updated based on https://github.com/horovod/horovod/blob/18cfedd5236885d9fd581ca53fe04a142486bb2b/horovod/tensorflow/__init__.py#L294
+class ElasticDistributedOptimizer(tf.train.Optimizer):
+    """An optimizer that wraps another tf.Optimizer, using an allreduce to
+    combine gradient values before applying gradients to model weights."""
+
+    def __init__(self, optimizer, name=None, use_locking=False, device_dense='',
+                device_sparse='', compression=hvd.Compression.none,
+                sparse_as_dense=False, op=hvd.Average, gradient_predivide_factor=1.0,
+                backward_passes_per_step=1, average_aggregated_gradients=False):
+        if name is None:
+            name = "Distributed{}".format(type(optimizer).__name__)
+        super(ElasticDistributedOptimizer, self).__init__(name=name, use_locking=use_locking)
+
+        self._optimizer = optimizer
+        self._allreduce_grads = hvd._make_allreduce_grads_fn(
+            name, device_dense, device_sparse, compression, sparse_as_dense, op,
+            gradient_predivide_factor)
+
+        self._agg_helper = None
+        if backward_passes_per_step > 1:
```

Comment:
In the elastic training, the `backward_passes_per_step` need to change. So, should we use the `agg_helper` default?

Reply:
Good idea, updated.